### ISSUE TITLE
Kernel panics in 5.15 kernels with the runtime hook

### ISSUE THREAD

**Issue by bahorn** [bahorn/skp#1]
## KASAN Issue

On 5.15 and probably before there is a crash that only occurs with KASAN:
```
[    5.573325] PATCHED KERNEL
[    5.573581] Called via UEFI Runtime hook
[    5.573900] general protection fault, probably for non-canonical address 0xdffffc0000000001: 0000 [#1] SMP KASAN NOPTI
[    5.574742] KASAN: null-ptr-deref in range [0x0000000000000008-0x000000000000000f]
[    5.575303] CPU: 0 PID: 1 Comm: init Not tainted 5.15.0-rc4+ #10
[    5.575752] Hardware name: QEMU Standard PC (i440FX + PIIX, 1996), BIOS
[    5.576259] RIP: 0010:pgd_free+0x146/0x260
[    5.576575] Code: 48 89 fa 48 c1 ea 03 80 3c 02 00 0f 85 f0 00 00 00 49 8d 7f 08 4c 8b 73 10 48 b8 00 00 00 00 00 fc ff df 48 89 fa 48 c1 ea 03 <80> 3c 02 00 0f 85 c3 00 00 00 4c 89 f2 4d 89 77 08 48 b8 00 00 00
[    5.577971] RSP: 0018:ffff888100a576d0 EFLAGS: 00010212
[    5.578360] RAX: dffffc0000000000 RBX: ffffea0004026380 RCX: ffffffffb2d8af0a
[    5.578883] RDX: 0000000000000001 RSI: 0000000000000000 RDI: 0000000000000008
[    5.579414] RBP: ffff888100a576f8 R08: 0000000000000000 R09: 0000000000000003
[    5.579937] R10: ffff88818098e000 R11: 0000000000000001 R12: ffff88810098e000
[    5.580462] R13: ffffffffb78167c0 R14: dead000000000122 R15: 0000000000000000
[    5.580998] FS:  000070217b67db48(0000) GS:ffff888106a00000(0000) knlGS:0000000000000000
[    5.581602] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[    5.582032] CR2: 00005f9b0a35e640 CR3: 0000000106a42000 CR4: 00000000000006f0
[    5.582563] Call Trace:
[    5.582752]  __mmdrop+0xd3/0x370
[    5.583007]  finish_task_switch.isra.0+0x479/0x6d0
[    5.583369]  __schedule+0x800/0x3960
[    5.583647]  ? try_to_wake_up+0x4d7/0x1390
[    5.583959]  ? io_schedule_timeout+0x160/0x160
[    5.584295]  ? sched_core_balance+0x890/0x890
[    5.584629]  __cond_resched+0x4a/0x80
[    5.584913]  __wait_for_common+0x97/0x350
[    5.585222]  ? usleep_range+0x150/0x150
[    5.585602]  ? bit_wait_timeout+0x170/0x170
[    5.585979]  ? __sanitizer_cov_trace_const_cmp8+0x18/0x20
[    5.586452]  wait_for_completion+0x24/0x30
[    5.586821]  virt_efi_get_variable+0x19a/0x1b0
[    5.587222]  efivar_entry_size+0xdb/0x170
[    5.587591]  efivarfs_callback+0x386/0x5b0
[    5.587960]  ? efivarfs_evict_inode+0x30/0x30
[    5.588350]  ? __kasan_check_write+0x14/0x20
[    5.588734]  ? _raw_spin_lock_irqsave+0x8e/0xf0
[    5.589134]  ? up+0x7a/0xc0
[    5.589405]  ? __sanitizer_cov_trace_const_cmp2+0x18/0x20
[    5.589876]  ? ucs2_strsize+0xeb/0x120
[    5.590222]  efivar_init+0x40b/0x850
[    5.590551]  ? efivarfs_evict_inode+0x30/0x30
[    5.590949]  ? efivar_variable_is_removable+0x320/0x320
[    5.591421]  ? __kasan_check_write+0x14/0x20
[    5.591828]  ? __sanitizer_cov_trace_const_cmp4+0x16/0x20
[    5.592303]  ? __d_instantiate+0x358/0x450
[    5.592623]  ? d_instantiate+0x87/0xa0
[    5.592919]  efivarfs_fill_super+0x25b/0x310
[    5.593246]  ? efivarfs_get_tree+0x30/0x30
[    5.593561]  ? efivarfs_get_tree+0x30/0x30
[    5.593871]  get_tree_single+0xd8/0x1d0
[    5.594171]  efivarfs_get_tree+0x22/0x30
[    5.594475]  vfs_get_tree+0x92/0x2f0
[    5.594778]  path_mount+0x6b9/0x1e00
[    5.595057]  ? putname+0x115/0x150
[    5.595326]  ? finish_automount+0x750/0x750
[    5.595798]  ? putname+0x11a/0x150
[    5.596298]  __x64_sys_mount+0x2ba/0x340
[    5.596869]  ? path_mount+0x1e00/0x1e00
[    5.597430]  ? exit_to_user_mode_prepare+0x38/0x1f0
[    5.598134]  do_syscall_64+0x5c/0x80
[    5.598660]  entry_SYSCALL_64_after_hwframe+0x44/0xae
[    5.599381] RIP: 0033:0x70217b62713f
[    5.599894] Code: 05 48 89 c7 e8 52 40 ff ff 5a c3 89 f6 50 b8 b0 00 00 00 0f 05 48 89 c7 e8 3e 40 ff ff 5a c3 49 89 ca 50 b8 a5 00 00 00 0f 05 <48> 89 c7 e8 29 40 ff ff 5a c3 31 f6 50 b8 a6 00 00 00 0f 05 48 89
[    5.602479] RSP: 002b:00007ffeed4400a0 EFLAGS: 00000246 ORIG_RAX: 00000000000000a5
[    5.603059] RAX: ffffffffffffffda RBX: 00007ffeed4400d0 RCX: 000070217b62713f
[    5.603575] RDX: 00005f9b0a35e206 RSI: 00005f9b0a35e20f RDI: 00005f9b0a35e221
[    5.604090] RBP: 00005f9b0a35e186 R08: 0000000000000000 R09: 00005f9b0a35e1eb
[    5.604605] R10: 000000000000040e R11: 0000000000000246 R12: 00005f9b0a35e25f
[    5.605122] R13: 0000000000000012 R14: 00005f9b0a35e26e R15: 00005f9b0a35e277
[    5.605643] Modules linked in:
[    5.605905] ---[ end trace fe24d575edde59ac ]---
[    5.606246] RIP: 0010:pgd_free+0x146/0x260
[    5.606563] Code: 48 89 fa 48 c1 ea 03 80 3c 02 00 0f 85 f0 00 00 00 49 8d 7f 08 4c 8b 73 10 48 b8 00 00 00 00 00 fc ff df 48 89 fa 48 c1 ea 03 <80> 3c 02 00 0f 85 c3 00 00 00 4c 89 f2 4d 89 77 08 48 b8 00 00 00
[    5.607892] RSP: 0018:ffff888100a576d0 EFLAGS: 00010212
[    5.608276] RAX: dffffc0000000000 RBX: ffffea0004026380 RCX: ffffffffb2d8af0a
[    5.608797] RDX: 0000000000000001 RSI: 0000000000000000 RDI: 0000000000000008
[    5.609327] RBP: ffff888100a576f8 R08: 0000000000000000 R09: 0000000000000003
[    5.609861] R10: ffff88818098e000 R11: 0000000000000001 R12: ffff88810098e000
[    5.610390] R13: ffffffffb78167c0 R14: dead000000000122 R15: 0000000000000000
[    5.610929] FS:  000070217b67db48(0000) GS:ffff888106a00000(0000) knlGS:0000000000000000
[    5.611536] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[    5.611961] CR2: 00005f9b0a35e640 CR3: 0000000106a42000 CR4: 00000000000006f0
[    5.612493] Kernel panic - not syncing: Attempted to kill init! exitcode=0x0000000b
[    5.613126] Kernel Offset: 0x31c00000 from 0xffffffff81000000 (relocation range: 0xffffffff80000000-0xffffffffbfffffff)
[    5.613939] ---[ end Kernel panic - not syncing: Attempted to kill init! exitcode=0x0000000b ]---
```

I git bisect'd this down to being fixed in this commit:
```
c597bfddc9e9e8a63817252b67c3ca0e544ace26 is the first bad commit
commit c597bfddc9e9e8a63817252b67c3ca0e544ace26
Author: Frederic Weisbecker <frederic@kernel.org>
Date:   Tue Sep 14 12:31:34 2021 +0200

    sched: Provide Kconfig support for default dynamic preempt mode

    Currently the boot defined preempt behaviour (aka dynamic preempt)
    selects full preemption by default when the "preempt=" boot parameter
    is omitted. However distros may rather want to default to either
    no preemption or voluntary preemption.

    To provide with this flexibility, make dynamic preemption a visible
    Kconfig option and adapt the preemption behaviour selected by the user
    to either static or dynamic preemption.

    Signed-off-by: Frederic Weisbecker <frederic@kernel.org>
    Signed-off-by: Peter Zijlstra (Intel) <peterz@infradead.org>
    Link: https://lkml.kernel.org/r/20210914103134.11309-1-frederic@kernel.org

 kernel/Kconfig.preempt | 32 +++++++++++++++++++++++---------
 kernel/sched/core.c    | 29 ++++++++++++++++++++++++++---
 2 files changed, 49 insertions(+), 12 deletions(-)
```

So it seems related to preempt behavior? I did try and play around with the `preempt` kernel cmdline option but couldn't get any different behavior.

This one doesn't seem to exist without KASAN being enabled, as otherwise the 5.15 kernel I just build on the commit just before that works fine.

## Ubuntu Kernels

There also seems to be another issue that shows up in the ubuntu kernel builds with:
```
[    0.564628] PATCHED KERNEL
[    0.564824] Called via UEFI Runtime hook
[    0.565144] relative relocation: 1885
[    0.565395] relative relocation: 1584
[    0.565641] relative relocation: 4656
[    0.565898] relative relocation: 1008
[    0.566924] relocating sym: register_ftrace_function
[    0.567711] relocating sym: commit_creds
[    0.568667] relocating sym: kallsyms_lookup_name
[    0.569484] relocating sym: prepare_creds
[    0.570615] relocating sym: ftrace_set_filter_ip
[    0.571771] relocating sym: unregister_ftrace_function
[    0.576582] relocating sym: _printk
[    0.576826] RX
[    0.576980] RW
[    0.577099] Entrypoint: ffffac2e800613f0
[    0.577362] Greetings from SHELF
[    0.578928] rootkit: Loaded >:-)
[    0.579153] ------------[ cut here ]------------
[    0.579469] [Firmware Bug]: Page fault caused by firmware at PA: 0xfffffffefda06991
[    0.580039] WARNING: CPU: 0 PID: 95 at arch/x86/platform/efi/quirks.c:735 efi_crash_gracefully_on_page_fault+0x5a/0xe0
[    0.580748] Modules linked in:
[    0.580964] CPU: 0 PID: 95 Comm: kworker/u2:3 Not tainted 5.15.0-117-generic #127-Ubuntu
[    0.581529] Hardware name: QEMU Standard PC (i440FX + PIIX, 1996), BIOS
[    0.582005] Workqueue: efi_rts_wq efi_call_rts
[    0.582335] RIP: 0010:efi_crash_gracefully_on_page_fault+0x5a/0xe0
[    0.582751] Code: 49 81 fc ff 0f 00 00 76 08 48 3d 10 a1 83 b6 74 0a 4c 8b 65 f8 c9 c3 cc cc cc cc 4c 89 e6 48 c7 c7 90 67 9d b4 e8 80 b1 c6 00 <0f> 0b 83 3d 0d 48 38 03 0a 0f 84 50 9e c6 00 e8 42 1c 00 00 e8 ad
[    0.583990] RSP: 0018:ffffac2e805d7ad0 EFLAGS: 00010086
[    0.584340] RAX: 0000000000000000 RBX: 0000000000000001 RCX: ffffffffb537b7c8
[    0.584868] RDX: 0000000000000000 RSI: 0000000000000003 RDI: 0000000000000001
[    0.585425] RBP: ffffac2e805d7ad8 R08: 0000000000000003 R09: 6166206567615020
[    0.585976] R10: 0000000061776d72 R11: 000000007269465b R12: fffffffefda06991
[    0.586529] R13: ffffac2e805d7c38 R14: 0000000000000010 R15: fffffffefda06991
[    0.587106] FS:  0000000000000000(0000) GS:ffff9c513bc00000(0000) knlGS:0000000000000000
[    0.587790] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[    0.588244] CR2: fffffffefda06991 CR3: 0000000025e10000 CR4: 00000000000006f0
[    0.588803] Call Trace:
[    0.589001]  <TASK>
[    0.589171]  ? show_trace_log_lvl+0x1d6/0x2ea
[    0.589520]  ? show_trace_log_lvl+0x1d6/0x2ea
[    0.589866]  ? page_fault_oops+0x4f/0x170
[    0.590202]  ? show_regs.part.0+0x23/0x29
[    0.590593]  ? show_regs.cold+0x8/0xd
[    0.590888]  ? efi_crash_gracefully_on_page_fault+0x5a/0xe0
[    0.591333]  ? __warn+0x8c/0x100
[    0.591600]  ? efi_crash_gracefully_on_page_fault+0x5a/0xe0
[    0.592044]  ? report_bug+0xa4/0xd0
[    0.592324]  ? handle_bug+0x39/0x90
[    0.592580]  ? exc_invalid_op+0x19/0x70
[    0.592856]  ? asm_exc_invalid_op+0x1b/0x20
[    0.593158]  ? efi_crash_gracefully_on_page_fault+0x5a/0xe0
[    0.593549]  page_fault_oops+0x4f/0x170
[    0.593828]  ? search_exception_tables+0x61/0x70
[    0.594163]  kernelmode_fixup_or_oops+0xa2/0x120
[    0.594495]  __bad_area_nosemaphore+0x15d/0x1a0
[    0.594817]  bad_area_nosemaphore+0x16/0x20
[    0.595120]  do_kern_addr_fault+0x62/0x80
[    0.595411]  exc_page_fault+0xe7/0x170
[    0.595691]  asm_exc_page_fault+0x27/0x30
[    0.595981] RIP: 0010:0xfffffffefda06991
[    0.596266] Code: Unable to access opcode bytes at RIP 0xfffffffefda06967.
[    0.596757] RSP: 0018:ffffac2e805d7ce8 EFLAGS: 00010246
[    0.597127] RAX: 0000000000000000 RBX: 0000000000000005 RCX: ffffffffb537b7c8
[    0.597629] RDX: 0000000000000000 RSI: 0000000000000003 RDI: 0000000000000001
[    0.598128] RBP: ffffac2e805d7cf0 R08: 0000000000000003 R09: 616f4c203a74696b
[    0.598625] R10: 20646564616f4c20 R11: 3a74696b746f6f72 R12: ffffffffb4a34958
[    0.599123] R13: ffffac2e8001fdb0 R14: 0000000000000000 R15: ffffac2e8001fd98
[    0.599661]  ? 0xffffffffb3400000
[    0.599956]  ? switch_mm_irqs_off+0x1da/0x550
[    0.600316]  ? __efi_call+0x28/0x30
[    0.600609]  ? switch_mm+0x20/0x40
[    0.600873]  ? efi_enter_mm+0x39/0x40
[    0.601147]  ? efi_call_rts+0x1a6/0x770
[    0.601431]  ? process_one_work+0x22b/0x3d0
[    0.601789]  ? worker_thread+0x53/0x420
[    0.602113]  ? process_one_work+0x3d0/0x3d0
[    0.602451]  ? kthread+0x12a/0x150
[    0.602713]  ? set_kthread_struct+0x50/0x50
[    0.603069]  ? ret_from_fork+0x22/0x30
[    0.603368]  </TASK>
[    0.603564] ---[ end trace dd0fbfc901f0fdca ]---
[    0.603942] efi: Froze efi_rts_wq and disabled EFI Runtime Services
[    0.604467] ------------[ cut here ]------------
[    0.604844] WARNING: CPU: 0 PID: 1 at drivers/firmware/efi/runtime-wrappers.c:311 virt_efi_get_variable+0x179/0x180
[    0.605655] Modules linked in:
[    0.605891] CPU: 0 PID: 1 Comm: swapper/0 Tainted: G        W         5.15.0-117-generic #127-Ubuntu
[    0.606591] Hardware name: QEMU Standard PC (i440FX + PIIX, 1996), BIOS
[    0.607124] RIP: 0010:virt_efi_get_variable+0x179/0x180
[    0.607558] Code: f5 24 2f 00 48 c7 c7 38 a1 83 b6 e8 71 56 31 00 48 8b 05 f2 a1 97 02 48 ba 15 00 00 00 00 00 00 80 48 39 d0 0f 85 06 ff ff ff <0f> 0b e9 ff fe ff ff 0f 1f 44 00 00 55 48 89 e5 41 56 49 89 d6 41
[    0.609096] RSP: 0018:ffffac2e8001fd40 EFLAGS: 00010246
[    0.609479] RAX: 8000000000000015 RBX: ffffac2e8001fd84 RCX: 0000000000000000
[    0.610016] RDX: 8000000000000015 RSI: ffffac2e8001fcd8 RDI: ffffffffb683a140
[    0.610536] RBP: ffffac2e8001fd70 R08: 0000000000000000 R09: 0000000000000000
[    0.611057] R10: 0000000000000001 R11: 0000000000000000 R12: ffffac2e8001fd98
[    0.611593] R13: 0000000000000000 R14: ffffac2e8001fdb0 R15: ffffffffb4a34958
[    0.612116] FS:  0000000000000000(0000) GS:ffff9c513bc00000(0000) knlGS:0000000000000000
[    0.612703] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[    0.613121] CR2: fffffffefda06967 CR3: 0000000025e10000 CR4: 00000000000006f0
[    0.613642] Call Trace:
[    0.613833]  <TASK>
[    0.613994]  ? show_trace_log_lvl+0x1d6/0x2ea
[    0.614316]  ? show_trace_log_lvl+0x1d6/0x2ea
[    0.614637]  ? load_uefi_certs+0xd6/0x259
[    0.614934]  ? show_regs.part.0+0x23/0x29
[    0.615231]  ? show_regs.cold+0x8/0xd
[    0.615507]  ? virt_efi_get_variable+0x179/0x180
[    0.615848]  ? __warn+0x8c/0x100
[    0.616090]  ? virt_efi_get_variable+0x179/0x180
[    0.616431]  ? report_bug+0xa4/0xd0
[    0.616690]  ? handle_bug+0x39/0x90
[    0.616952]  ? exc_invalid_op+0x19/0x70
[    0.617236]  ? asm_exc_invalid_op+0x1b/0x20
[    0.617545]  ? virt_efi_get_variable+0x179/0x180
[    0.617893]  load_uefi_certs+0xd6/0x259
[    0.618176]  ? load_moklist_certs+0x1a0/0x1a0
[    0.618496]  do_one_initcall+0x49/0x1e0
[    0.618779]  do_initcalls+0x12f/0x159
[    0.619050]  kernel_init_freeable+0x162/0x1b5
[    0.619372]  ? rest_init+0x100/0x100
[    0.619659]  kernel_init+0x1b/0x150
[    0.619927]  ? rest_init+0x100/0x100
[    0.620209]  ret_from_fork+0x22/0x30
[    0.620479]  </TASK>
[    0.620646] ---[ end trace dd0fbfc901f0fdcb ]---
[    0.621004] efi: EFI Runtime Services are disabled!
```
Where it seems the stack is getting misaligned, but only after my calls to setup the payloads ftrace hook.
Seem to be able to reproduce this on a custom kernel by taking Ubuntu's config, bisected down to same commit as above.

**Comment by bahorn** [comment]
Seems `CONFIG_PREEMPT` has no issues, `CONFIG_PREEMPT_NONE` has the page fault but continues execution and `CONFIG_PREEMPT_VOLUNTARY` kills the kernel.

It does not seem to occur with `CONFIG_PREEMPT_DYNAMIC`, though it gets disabled if you have anything other than `CONFIG_PREEMPT` enabled?

**Comment by bahorn** [comment]
Writing my own version of `preempt_disable()` and `preempt_enable()` seems to have fixed the issue. Just need to make it work across kernel versions as `__preempt_count` got moved into the `pcpu_hot` struct in recent kernels.

**Comment by bahorn** [comment]
Should be now fixed with that commit.
